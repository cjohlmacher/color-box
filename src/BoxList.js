import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import './BoxList.css';

const BoxList = () => {
    const [boxes,setBoxes] = useState([]);
    const boxComponents = boxes.map(({id, backgroundColor, width, height}) => {
        return <Box key={id} deleteBox={() => deleteBox(id)} backgroundColor={backgroundColor} width={width} height={height} />
    });

    const deleteBox = (id) => {
        setBoxes(boxes.filter((box) => box.id !== id ));
    };

    const addBox = (newBox) => {
        const id = uuid();
        setBoxes([...boxes, {id: id, backgroundColor: newBox.backgroundColor, width: newBox.width, height: newBox.height}]);
    };

    return (
        <div className="BoxList">
            <ul>
                {boxComponents}
            </ul>
            <NewBoxForm addBox={addBox} />
        </div>
    )
};

export default BoxList;