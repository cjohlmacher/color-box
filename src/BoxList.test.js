import { fireEvent, render } from "@testing-library/react";
import BoxList from './BoxList';

const addTestBox = (getByLabelText, getByRole) => {
    const colorInput = getByLabelText('Background Color:');
    const widthInput = getByLabelText('Width:');
    const heightInput = getByLabelText('Height:');
    const submitButton = getByRole('button');
    fireEvent.change(colorInput,{target: {value: "aqua", name: "backgroundColor"}});
    fireEvent.change(widthInput,{target: {value: 30, name: "width"}});
    fireEvent.change(heightInput,{target: {value: 50, name: "height"}});
    fireEvent.click(submitButton);
}

it('renders without crashing', () => {
    render(<BoxList />);
});

it('matches snapshot', () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('adds a box when the form is submitted', () => {
    const { getByLabelText, getByRole, getByTestId, asFragment } = render(<BoxList />);
    addTestBox(getByLabelText, getByRole);
    const colorBox = getByTestId('box');
    expect(colorBox).toHaveStyle('background-color: aqua');
    expect(colorBox).toHaveStyle('width: 30px');
    expect(colorBox).toHaveStyle('height: 50px');
    expect(colorBox).toBeInTheDocument();

});

it('deletes a box when a delete button is clicked', () => {
    const { getByLabelText, getByRole, getByTestId } = render(<BoxList />);
    addTestBox(getByLabelText, getByRole);
    const colorBox = getByTestId('box');
    const deleteButton = getByRole('button', {name: 'X'});
    fireEvent.click(deleteButton);
    expect(colorBox).not.toBeInTheDocument();
})