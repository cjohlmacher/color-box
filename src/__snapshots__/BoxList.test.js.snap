// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <ul />
    <form>
      <label
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
      <button
        type="submit"
      >
        Add Box!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
